src/coprocessor/endpoint.rs

```diff
@@ -342,13 +348,32 @@ impl<E: Engine> Endpoint<E> {
             self.batch_row_limit
         }
     }
+
     #[inline]
     fn async_snapshot(
         engine: &E,
-        ctx: &kvrpcpb::Context,
+        ctx: &ReqContext,
     ) -> impl std::future::Future<Output = Result<E::Snap>> {
-        kv::snapshot(engine, None, ctx).map_err(Error::from)
+        let mut snap_ctx = SnapContext {
+            pb_ctx: &ctx.context,
+            ..Default::default()
+        };
+        // need to pass start_ts and ranges to check memory locks for replica read
+        if ctx.context.get_replica_read() && ctx.context.get_isolation_level() == IsolationLevel::Si
```

Comment:
Can function `need_check_locks_in_replica_read` be used here?

Reply:
Ah.. Yes.